text/0193-fixing-type-unions.md

```diff
@@ -0,0 +1,581 @@
+a list of all properties found in the CloudFormation Resource Specification at the time of that release, and branching code-generation based on
+whether the property is in the list or not.
+
+In order to gradually enahnce the developer experience in statically typed languages, a new `awslint` rule will be introduced to prohibit introduction
+of new inline _type unions_. Exceptions will be added for all existing APIs.
+
+## Generating _union-like_ types
+
+Code generation needs to be changed in `jsii-pacmak` so that languages without native support for _type unions_ (**C#**, **Go**, and **Java**) have
+additional _union-like_ types emitted. The proposed API for a union such as `export type UnionType = Foo | Map<Bar> | Baz[]` in each currently
+supported statically typed language is the following (subject ot change):
+
+- In **C#**
+
+  ```csharp
+  // UnionType.cs
+  public sealed class UnionType {
+    public static UnionType FromFoo(Foo value) { /* ... */ }
+    public static UnionType FromMapOfBar(IReadOnlyDictionary<Bar> value) { /* ... */ }
+    public static UnionType FromListOfBaz(IReadOnlyList<Baz> value) { /* ... */ }
+
+    public Foo? Foo { get; }
+    public IReadOnlyDictionary<Bar>? MapOfBar { get; }
+    public IReadOnlyList<Baz>? ListOfBaz { get; }
+
+    // Private constructor ensures all initialization happens though the factories
+    private UnionType(/*...*/) { /*...*/ }
+  }
+  ```
+
+- In **Go**:
+
+  > :construction: TBD
+
+- In **Java**:
+
+  ```java
+  // UnionType.java
+  public abstract class UnionType {
+    public static UnionType fromFoo(Foo value) {
+      return new UnionType.FooUnionType(value);
+    }
+    public static UnionType fromMapOfBar(Map<String, Bar> value) {
+      return new UnionType.MapOfBarUnionType(value);
+    }
+    public static UnionType fromListOfBaz(Baz... value) {
+      return fromListOfBaz(Arrays.asList(value));
+    }
+    public static UnionType fromListOfBaz(List<Baz> value) {
+      return new UnionType.ListOfBaz(value);
+    }
+
+    public @Nullable Foo asFoo() { return null; }
+    public @Nullable Map<String, Bar> asMapOfBar() { return null; }
+    public @Nullable List<Baz> asListOfBaz() { return null; }
+
+    // Private constructor ensures all initialization happens through the factories
+    private UnionType(/* ... */) { /* ... */ }
+
+    private static final class FooUnionType extends UnionType { /*...*/ }
+    private static final class MapOfBarUnionType extends UnionType { /*...*/ }
+    private static final class ListOfBazUnionType extends UnionType { /*...*/ }
+  }
+  ```
+
+The _jsii Kernel_ will need adjustments in order to better support those types. The generated libraries will provide a new argument to the _jsii
+kernel_'s `load` API if it's code was generated with `--union-like-types` enabled. This way, the kernel will be bale to reference the declared _type
```

Comment:
I'd still like to see what the TypeScript code that produces a `UnionType` looks like, and a description of how the kernel uses that.

As far as I can tell, you're still stuck with the:

```ts
export function banana(): UnionType {
  return {
    fu: 'manchu'
  };
}
```

Problem of needing to do field analysis on object literals.

Reply:
No - we'd wire the value as a `UnionType` typed. So we're not actually promising any more than what the declared type tells us.

Reply:
Okay. But then what does a Java receiver do? How will it decide between `FuType | BarType` in this instance?